### PR TITLE
Refactor reproduction architecture and shared utilities

### LINKED ISSUE #60: [Roadmap] Scientific reproduction notebooks and enabling XDRL work
Source paper list: #48

## Notebook deliverables

Each paper has one dedicated notebook issue. These are documentation/scientific-work deliverables, not XDRL feature issues.

- [ ] #56 — functional neural modules in MiniGrid
- [ ] #55 — emergent planning in Sokoban
- [ ] #58 — goal representations and steering in Procgen Maze
- [ ] #57 — NA2Q on one cooperative MARL task
- [ ] #59 — expert-concept policies on one simulated MARL task

Every notebook lives under `docs/source/reproductions/`, separate from the tutorial notebooks, and reports source revisions, controls, provenance, result artifacts, deviations, and claim limits.

## Enabling XDRL work

These issues change documentation design or the reusable library contract. They do not implement a paper.

- [ ] #52 — **Docs:** add a scientific reproductions gallery separate from tutorials
- [ ] #53 — **Feature:** record external inputs and result artifacts in workflow provenance
- [ ] #54 — **Feature:** run paired TDHook workflows with matched XDRL provenance
- [ ] #61 — **Design:** represent repeated internal computation in recurrent interactions
- [ ] #62 — **Design:** type coalition terms and semantic masks in multi-agent value decomposition

## Boundaries

- TorchRL/TensorDict retain model, environment, trainer, and data ownership.
- TDHook retains generic probing, capture, intervention, attribution, and hook-target mechanics.
- XDRL changes are limited to typed RL interaction semantics, validated execution, lifecycle, and provenance required across reproductions.
- Notebook smoke execution is maintenance evidence, not a reproduced scientific result.
- A new library gap discovered during a notebook reproduction must become a focused design/feature issue rather than notebook-local framework code.

### PR DESCRIPTION
**Pull request by Xmaster6y**
## Summary

- separate declarative RL semantics from runtime interaction orchestration with one canonical module boundary
- centralize deterministic tensor, module, byte, and repository identities used across reproductions
- remove duplicated notebook infrastructure and standardize the reproduction narrative around methods, results, and scope
- remove gate/checklist-style scaffolding, stale outputs, and execution metadata from the reproduction notebooks

## Validation

- `uv run pre-commit run --all-files`
- `uv run pytest tests --cov=src --cov-report=term-missing --cov-fail-under=50 -q` (295 passed, 95.86% coverage)
- `uv run --group docs sphinx-build -W --keep-going -b html docs/source docs/build/html`
- all changed notebooks executed successfully from fresh kernels

Closes #60

### CODE REVIEW

File: src/xdrl/digests.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+def tensor_digest(tensor: torch.Tensor) -> str:
+    """Hash a tensor's dtype, shape, and contiguous CPU values."""
+    value = tensor.detach().cpu().contiguous()
+    header = json.dumps(
+        {"dtype": str(value.dtype), "shape": list(value.shape)},
+        sort_keys=True,
+        separators=(",", ":"),
+    ).encode()
+    return bytes_digest(header + value.numpy().tobytes())
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply src/xdrl/digests.py]: Fixed in 6a6ff1c. `tensor_digest` now hashes a contiguous tensor through its raw `uint8` storage view, so dtypes without NumPy conversion support (including `bfloat16`) are handled. Added a bfloat16 regression test.

File: src/xdrl/digests.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        {"name": name, "sha256": tensor_digest(tensor)}
+        for name, tensor in sorted(named_tensors, key=lambda item: item[0])
+    ]
+    payload = json.dumps(manifest, sort_keys=True, separators=(",", ":")).encode()
+    return bytes_digest(payload)
+
+
+def module_digest(module: torch.nn.Module) -> str:
+    """Hash a module's named parameters and buffers deterministically."""
+    return named_tensor_digest(module.state_dict().items())
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply src/xdrl/digests.py]: Fixed in 6a6ff1c. `module_digest` now filters `state_dict()` to tensor entries, matching its parameters-and-buffers contract and safely ignoring non-tensor `_extra_state`. Added a module regression test covering extra-state mutation and parameter mutation.

File: src/xdrl/interactions.py [hunk 3]
```diff
@@ -8,438 +8,71 @@
...
-from typing import TYPE_CHECKING, Any, Callable, Iterator, Protocol
+from typing import TYPE_CHECKING, Any, Protocol
 
 import torch
 from tensordict import TensorDictBase
 from tensordict.nn import TensorDictModuleBase
 from torchrl.envs.utils import set_exploration_type
 
 from xdrl.interventions import InterventionTiming
+from xdrl.semantics import (
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply src/xdrl/interactions.py]: Fixed in 6a6ff1c. `AgentIdentity` and `InternalCoordinate` are re-exported from `xdrl.interactions` and included in its `__all__`; a compatibility import regression test protects the established import path.

File: src/xdrl/digests.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply src/xdrl/digests.py]: Fixed in 6a6ff1c. The digest uses raw contiguous `uint8` storage bytes rather than dtype-dependent NumPy conversion, with explicit bfloat16 coverage.

File: src/xdrl/semantics.py [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+            if len(self.agent_identities) != self.n_agents:
+                raise ValueError("agent_identities must match the declared group size")
+            if len(set(self.agent_identities)) != len(self.agent_identities):
+                raise ValueError("agent_identities must be unique")
+            if any(type(agent) not in {str, int} or agent == "" for agent in self.agent_identities):
+                raise TypeError("agent identities must be non-empty strings or integers")
+        if self.agent_identities:
+            declared = set(self.agent_identities)
+            for agent in self.target.selector.agents:
+                if agent not in declared:
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply src/xdrl/semantics.py]: Fixed in 6a6ff1c. Explicit agent selection now compares both exact type and value, so `True` no longer aliases integer identity `1`. Added a regression test for a boolean selector against integer identities.

File: docs/source/reproductions/na2q-value-decomposition.ipynb [hunk 5]
```diff
@@ -95,6 +77,7 @@
     "    SemanticTarget,\n",
     "    TDHookWorkflowRunner,\n",
     "    TensorDictSchema,\n",
+    "    tensor_digest,\n",
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply docs/source/reproductions/na2q-value-decomposition.ipynb]: Fixed in 6a6ff1c. Restored the historical JSON header serialization (default separators) while retaining raw-byte tensor hashing. An exact known-digest regression test locks the existing reproduction-bundle wire format, and all reproduction notebooks—including NA2Q—execute successfully.

File: src/xdrl/digests.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply src/xdrl/digests.py]: Fixed in 6a6ff1c. Non-tensor state-dict values such as `_extra_state` are excluded from `module_digest`; tensor parameters and buffers remain deterministically hashed. Covered by a custom module regression test.

File: src/xdrl/semantics.py [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+    truncated_window: int | None = None
+    collector_mode: RecurrentCollectorMode = RecurrentCollectorMode.DIRECT
+
+    def __post_init__(self) -> None:
+        if not self.transitions:
+            raise ValueError("recurrent semantics require at least one state transition")
+        if self.burn_in < 0:
+            raise ValueError("burn_in must be non-negative")
+        if self.truncated_window is not None and self.truncated_window < 1:
+            raise ValueError("truncated_window must be positive")
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply src/xdrl/semantics.py]: Fixed in 6a6ff1c. `burn_in` and `truncated_window` now require exact `int` values (excluding bool) in addition to their range constraints. Regression cases cover booleans and floats for both fields.

File: src/xdrl/semantics.py [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+    topology: InteractionTopology
+    group: str
+    n_agents: int
+    target: SemanticTarget
+    agent_identities: tuple[AgentIdentity, ...] = ()
+
+    def __post_init__(self) -> None:
+        if not self.group:
+            raise ValueError("multi-agent group must be non-empty")
+        if self.n_agents < 1:
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply src/xdrl/semantics.py]: Fixed in 6a6ff1c. `n_agents` now requires an exact positive `int`, rejecting both booleans and floats. Both cases are covered by regression tests.

File: src/xdrl/semantics.py [hunk 8]
```diff
@@ -0,0 +1,430 @@
+"""Declarative semantics for typed reinforcement-learning interactions.
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply src/xdrl/semantics.py]: Fixed in 6a6ff1c. `xdrl.semantics` now defines an explicit public `__all__`, including the two public type aliases and excluding private helpers. Added export-surface coverage.

File: tests/unit/test_digests.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
+import torch
+
+from xdrl.digests import bytes_digest, module_digest, named_tensor_digest, tensor_digest
+
+
+def test_tensor_digest_tracks_metadata_and_values() -> None:
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply tests/unit/test_digests.py]: Fixed in 6a6ff1c. Added the requested same-dtype/same-shape/different-value assertion, alongside legacy-wire-format and bfloat16 digest regressions.

File: src/xdrl/interactions.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply src/xdrl/interactions.py]: Follow-up in ad7b0bd: after maintainer review, we intentionally removed these compatibility re-exports. `AgentIdentity` and `InternalCoordinate` now have one canonical home in `xdrl.semantics`, keeping the module boundary explicit and avoiding a permanent alias layer.

File: docs/source/reproductions/na2q-value-decomposition.ipynb (hunk 5 above)
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply docs/source/reproductions/na2q-value-decomposition.ipynb]: Follow-up in ad7b0bd: after maintainer review, we intentionally dropped legacy digest-wire compatibility. `tensor_digest` now has one compact deterministic encoding; paper bundles are revision-pinned and must be produced for the declared XDRL revision rather than migrated implicitly.
